src/vs/editor/contrib/codeAction/browser/codeActionMenu.ts

```diff
@@ -50,34 +63,297 @@ export interface CodeActionShowOptions {
+	private readonly _onDidHideContextMenu = new Emitter<void>();
+	private codeActionList!: List<ICodeActionMenuItem>;
+	private options: ICodeActionMenuItem[] = [];
+	private _visible: boolean = false;
+	readonly onDidHideContextMenu = this._onDidHideContextMenu.event;
+	private _ctxMenuWidgetVisible!: IContextKey<boolean>;
+	private viewItems: ICodeActionMenuItem[] = [];
+	private focusedEnabledItem!: number;
+	private currSelectedItem!: number;
+
+	public static readonly ID: string = 'editor.contrib.codeActionMenu';
+
+	public static get(editor: ICodeEditor): CodeActionMenu | null {
+		return editor.getContribution<CodeActionMenu>(CodeActionMenu.ID);
+	}
 
 	private readonly _keybindingResolver: CodeActionKeybindingResolver;
+	listRenderer: any;
 
 	constructor(
 		private readonly _editor: ICodeEditor,
 		private readonly _delegate: CodeActionWidgetDelegate,
 		@IContextMenuService private readonly _contextMenuService: IContextMenuService,
 		@IKeybindingService keybindingService: IKeybindingService,
 		@ILanguageFeaturesService private readonly _languageFeaturesService: ILanguageFeaturesService,
-		@ITelemetryService private readonly _telemetryService: ITelemetryService
+		@ITelemetryService private readonly _telemetryService: ITelemetryService,
+		@IThemeService _themeService: IThemeService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IContextViewService private readonly _contextViewService: IContextViewService,
+		@IContextKeyService private readonly _contextKeyService: IContextKeyService,
 	) {
 		super();
 
+		this.editor = _editor;
 		this._keybindingResolver = new CodeActionKeybindingResolver({
 			getKeybindings: () => keybindingService.getKeybindings()
 		});
+
+		this._ctxMenuWidgetVisible = Context.Visible.bindTo(_contextKeyService);
 	}
 
 	get isVisible(): boolean {
 		return this._visible;
 	}
 
+	private isCodeActionWidgetEnabled(model: ITextModel): boolean {
+		return this._configurationService.getValue('editor.econtrib.codeAction.enabled', {
+			resource: model.uri
+		});
+	}
+
+	private _onListSelection(e: IListEvent<ICodeActionMenuItem>): void {
+		if (e.elements.length) {
+			e.elements.forEach(element => {
+				if (element.isEnabled) {
+					element.action.run();
+				}
+			});
+			this.dispose();
+		}
+	}
+
+	private _onListFocus(e: IListEvent<ICodeActionMenuItem>): void {
+		this._ctxMenuWidgetIsFocused?.set(true);
+	}
+
+	private renderCodeActionMenuList(element: HTMLElement, inputArray: IAction[]): IDisposable {
+		const renderDisposables = new DisposableStore();
+		const renderMenu = document.createElement('div');
+		this.listRenderer = new CodeMenuRenderer();
+
+		const height = inputArray.length * 27;
+		renderMenu.style.height = String(height) + 'px';
+
+		renderMenu.id = 'codeActionMenuWidget';
+		renderMenu.classList.add('codeActionMenuWidget');
+
+		element.appendChild(renderMenu);
+
+		this.codeActionList = new List('codeActionWidget', renderMenu, {
```

Comment:
Since `codeActionList` is disposable, you can store it in a `MutableDisposable` to make sure that it isn't leaked when changing values